### DISCUSSION TITLE
QC for Inclusion/Exclusion Assessment (IE)

### DISCUSSION THREAD

**Discussion by mattroumaya** [Gilead-Public/gsm#205] [General]
# QC for Inclusion/Exclusion Assessment (IE)

- [ ] Documentation
    - [x] Function name captured in [Roxygen title](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block)(e.g. "Adverse Event Assessment")
    - [ ] Assessment Purpose captured in [Roxygen description](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."
    - [x] Input Data Requirements are described in the the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Data section (`#' @details # Data Requirements`).
    - [ ] Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the [Roxygen details](https://cran.r-project.org/web/packages/roxygen2/vignettes/rd.html#the-description-block) under a Statistics section (`#' @details # Statistical Assumptions`).
    - [ ] Parameters have detailed documentation using `@param`. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)
    - [x] All external dependencies are captured. Use `@importFrom` when importing 3 or fewer functions, and use `@import` otherwise. 
    - [x] Output Data Standards are described under `@returns`
    - [x] At least 1 example is provided under `@examples`
- [ ] Error Checking
    - [ ]  Basic checks for all all parameters should be included using `stopifnot()` or similar logic. (e.g. `stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput))`)
    - [ ] Tests confirm that `stopifnot()` parameter checks are working as expected. 
    - [ ] Tests confirm that the input data has required columns (if any)
    - [x] Tests confirm that the output data has the expected columns
    - [ ] Tests confirm intended functionality for each parameter
    - [ ] Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed. 
- [ ] Basic QC
    - [ ] Assessment has User Requirements + Qualification tests captured using {valtools} framework. Report is generating as expected and all checks are passing. 
    - [x] Code is well commented and easy to read
    - [x] No file paths or other company-specific data is present
    - [x] Function calls from non-tidyverse dependencies are called via `::`
    - [x] devtools::check() passes with no errors/warnings/notes
    - [x] package docs up to date. devtools::document() doesn't change any files
    - [x] Uses tidyverse best practices for standard data manipulation (start discussion thread if unclear)
    - [x] All new dependencies add significant value (start discussion thread)
    - [x] All GitHub Actions run with no errors

**Comment by mattroumaya** [comment]
## IE_Map_Raw()

**Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
- Minor, but [add commas to specify required columns](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Map_Raw.R#L22)
  - Suggested update: _ie dataset with required columns SUBJID, IVID, IECAT, IETESTCD, IEORRES_

**Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
- [Args to `IE_Map_Raw()`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Map_Raw.R#L36-L37) are both `= NULL` - remove to be consistent with other `Map*` functions
- Need checks for [`vExpected`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Map_Raw.R#L38)
  - Suggested: check for `is.character()` and that `length == 2`
- Need checks for `NA` handling
- Need checks for duplicate ID handling

**Tests confirm that stopifnot() parameter checks are working as expected.**
- Can combine [lines 12 - 24](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_IE_Map_Raw.R#L12-L24)
- Need tests for `vExpected`
- Need tests for `NA` handling
- Need tests for duplicate ID handling

**Tests confirm that the input data has required columns (if any)**
- Need test for `SubjectID` in `dfRDSL`
- Suggest adding the following [here](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_IE_Map_Raw.R#L41):

```r
  expect_error(
    IE_Map_Raw(
      clindata::raw_ie_a2 ,
      clindata::rawplus_rdsl%>% select(-SubjectID)
    )
  )
```

**Tests confirm intended functionality for each parameter**
- Need tests for `vExpected`

**Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
- Need checks for `NA` handling

---

## IE_Assess()

**Assessment Purpose captured in Roxygen description (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."**
- Purpose contains **TODO**:  _The Inclusion/Exclusion Assessment uses the standard GSM data pipeline **(TODO add link to data vignette)** to flag sites with Inclusion / Exclusion irregularities. More details regarding the data pipeline and statistical methods are described below._

**Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the Roxygen details under a Statistics section (#' @details # Statistical Assumptions).**
- There is no statistics section - unsure if needed though, because only doing flagging with thresholds

**Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
- nThreshold description should probably be updated - is worded somewhat awkwardly ("nThreshold integer threshold values Flagging, integer values greater than will be flagged.")

**Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
- Need check for [`nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Assess.R#L44-L49)
  - Check if `is.numeric()`
  - Check if `length == 1`

**Tests confirm that stopifnot() parameter checks are working as expected.**
- Need checks for `nThreshold`
- Need checks for `NA` handling
- Need checks for duplicate `ID` handling

**Tests confirm intended functionality for each parameter**
- Need tests for `nThreshold`

**Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
- Need checks for `NA` handling


@jwildfire ready for your review

**Reply by jwildfire** [reply]
> ## IE_Map_Raw()
> **Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
> 
> * Minor, but [add commas to specify required columns](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Map_Raw.R#L22)
>
>   * Suggested update: _ie dataset with required columns SUBJID, IVID, IECAT, IETESTCD, IEORRES_
 
👍 
 
> **Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
> 
> * [Args to `IE_Map_Raw()`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Map_Raw.R#L36-L37) are both `= NULL` - remove to be consistent with other `Map*` functions

 👍 

> * Need checks for [`vExpected`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Map_Raw.R#L38)
  
 👍 

>   * Suggested: check for `is.character()` and that `length == 2`

 👍 

> * Need checks for `NA` handling

👍 . add for IE, but ignore for RDSL for now. 

> * Need checks for duplicate ID handling
Will handle in #159 

> **Tests confirm that stopifnot() parameter checks are working as expected.**
> 
> * Can combine [lines 12 - 24](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_IE_Map_Raw.R#L12-L24)
> * Need tests for `vExpected`

👍 

> * Need tests for `NA` handling

👍  dfIE only

> * Need tests for duplicate ID handling
Skip for now. Will handle in #159. 
 
> **Tests confirm that the input data has required columns (if any)**
> 
> * Need test for `SubjectID` in `dfRDSL`
 Skip for now. Will handle in #159. 

> * Suggest adding the following [here](https://github.com/Gilead-BioStats/gsm/blob/dev/tests/testthat/test_IE_Map_Raw.R#L41):
> 
> ```r
>   expect_error(
>     IE_Map_Raw(
>       clindata::raw_ie_a2 ,
>       clindata::rawplus_rdsl%>% select(-SubjectID)
>     )
>   )
> ```
> 
> **Tests confirm intended functionality for each parameter**
> 
> * Need tests for `vExpected`
👍  

> **Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
> 
> * Need checks for `NA` handling
👍  dfIE only.

**Reply by jwildfire** [reply]
> ## IE_Assess()
> **Assessment Purpose captured in Roxygen description (e.g. "Evaluates site-level level Adverse Event Rates and flags rates that are abnormally high or low."**
> 
> * Purpose contains **TODO**:  _The Inclusion/Exclusion Assessment uses the standard GSM data pipeline **(TODO add link to data vignette)** to flag sites with Inclusion / Exclusion irregularities. More details regarding the data pipeline and statistical methods are described below._

Ok for now. Will update with #83 

> **Statistical Assumptions - (Assess() and Analyze() only) List the statistical methods used and refer to detailed description in Analyze function in a the Roxygen details under a Statistics section (#' @details # Statistical Assumptions).**
> 
> * There is no statistics section - unsure if needed though, because only doing flagging with thresholds

👍  Yes, we should add a section with a brief description. 

> **Parameters have detailed documentation using @param. Each parameter should include: parameter name, brief description, usage details, the default value (if any), is it required, list of valid options (if applicable)**
> 
> * nThreshold description should probably be updated - is worded somewhat awkwardly ("nThreshold integer threshold values Flagging, integer values greater than will be flagged.")
> 

👍 

> **Basic checks for all all parameters should be included using stopifnot() or similar logic. (e.g. stopifnot("dfInput is not a dataframe"=is.data.frame(dfInput)))**
> 
> * Need check for [`nThreshold`](https://github.com/Gilead-BioStats/gsm/blob/dev/R/IE_Assess.R#L44-L49)
>   
>   * Check if `is.numeric()`
>   * Check if `length == 1`

👍 

> **Tests confirm that stopifnot() parameter checks are working as expected.**
> 
> * Need checks for `nThreshold`
> * Need checks for `NA` handling
> * Need checks for duplicate `ID` handling

👍  
 
> **Tests confirm intended functionality for each parameter**
> 
> * Need tests for `nThreshold`

👍 

> **Tests confirm that missing data in required columns is handled appropriately and errors/warnings are produced if needed.**
> * Need checks for `NA` handling

👍

**Comment by jwildfire** [comment]
Looks good @mattroumaya. OK to file issues in v0.2
